src/provider/shplonk.rs

```diff
@@ -0,0 +1,660 @@
+use crate::provider::kzg_commitment::KZGCommitmentEngine;
+use crate::provider::non_hiding_kzg::{
+  trim, KZGProverKey, KZGVerifierKey, UVKZGPoly, UniversalKZGParam,
+};
+use crate::provider::pedersen::Commitment;
+use crate::provider::traits::DlogGroup;
+use crate::provider::util::iterators::DoubleEndedIteratorExt;
+use crate::spartan::polys::univariate::UniPoly;
+use crate::traits::commitment::Len;
+use crate::traits::evaluation::EvaluationEngineTrait;
+use crate::traits::{Engine as NovaEngine, Group, TranscriptEngineTrait, TranscriptReprTrait};
+use crate::{CommitmentEngineTrait, NovaError};
+use ff::{Field, PrimeFieldBits};
+use group::{Curve, Group as group_Group};
+use pairing::{Engine, MillerLoopResult, MultiMillerLoop};
+use rayon::iter::{
+  IndexedParallelIterator, IntoParallelIterator, IntoParallelRefMutIterator, ParallelIterator,
+};
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use std::marker::PhantomData;
+
+use crate::provider::hyperkzg::EvaluationEngine as HyperKZG;
+use itertools::Itertools;
+use ref_cast::RefCast as _;
+use std::sync::Arc;
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(bound(
+  serialize = "E::G1Affine: Serialize, E::Fr: Serialize",
+  deserialize = "E::G1Affine: Deserialize<'de>, E::Fr: Deserialize<'de>"
+))]
+pub struct EvaluationArgument<E: Engine> {
+  comms: Vec<E::G1Affine>,
+  evals: Vec<Vec<E::Fr>>,
+  R_x: Vec<E::Fr>,
+  C_Q: E::G1Affine,
+  C_H: E::G1Affine,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct EvaluationEngine<E, NE> {
+  _p: PhantomData<(E, NE)>,
+}
+
+impl<E, NE> EvaluationEngine<E, NE>
+where
+  E: Engine,
+  NE: NovaEngine<GE = E::G1, Scalar = E::Fr, CE = KZGCommitmentEngine<E>>,
+  E::Fr: Serialize + DeserializeOwned,
+  E::G1Affine: Serialize + DeserializeOwned,
+  E::G2Affine: Serialize + DeserializeOwned,
+  E::G1: DlogGroup<ScalarExt = E::Fr, AffineExt = E::G1Affine>,
+  E::Fr: PrimeFieldBits,
+  <E::G1 as Group>::Base: TranscriptReprTrait<E::G1>,
+  E::Fr: TranscriptReprTrait<E::G1>,
+  E::G1Affine: TranscriptReprTrait<E::G1>,
+{
+  fn compute_a(c_q: &E::G1Affine, transcript: &mut impl TranscriptEngineTrait<NE>) -> E::Fr {
+    transcript.absorb(b"C_Q", c_q);
+    transcript.squeeze(b"a").unwrap()
+  }
+
+  fn compute_pi_polynomials(hat_P: &[E::Fr], point: &[E::Fr], eval: &E::Fr) -> Vec<Vec<E::Fr>> {
+    let mut polys: Vec<Vec<E::Fr>> = Vec::new();
+    polys.push(hat_P.to_vec());
+
+    for i in 0..point.len() - 1 {
+      let Pi_len = polys[i].len() / 2;
+      let mut Pi = vec![E::Fr::ZERO; Pi_len];
+
+      #[allow(clippy::needless_range_loop)]
+      Pi.par_iter_mut().enumerate().for_each(|(j, Pi_j)| {
+        *Pi_j =
+          point[point.len() - i - 1] * (polys[i][2 * j + 1] - polys[i][2 * j]) + polys[i][2 * j];
+      });
```

Comment:
aka:
```rust
      (0..Pi_len).into_par_iter().map(|j| {
        point[ell - i - 1] * (polys[i][2 * j + 1] - polys[i][2 * j]) + polys[i][2 * j]
      }).collect_into_vec(&mut Pi);
```

Reply:
Thanks. Indeed cleaner and doesn't require adding clippy exception